src/libYARP_OS/src/Node.cpp

```diff
@@ -213,122 +241,170 @@ class yarp::os::Node::Helper : public PortReader
 
     virtual bool read(ConnectionReader& reader);
 
-    Contact lookup(const ConstString& topic)
-    {
-        mutex.lock();
-        std::map<ConstString,NodeItem>::const_iterator i =
-            by_part_name.find(topic);
-        if (i == by_part_name.end()) {
-            mutex.unlock();
-            return Contact();
-        }
-        Contact c = i->second.contactable->where();
-        mutex.unlock();
-        return c;
-    }
-
     void getBusStats(NodeArgs& na)
     {
-        na.reply.addList();
+        na.reply = Value();
         na.success();
     }
 
     void getBusInfo(NodeArgs& na)
     {
-        na.reply.addList();
+        unsigned int opaque_id = 1;
+        ROSReport report;
+        Value v;
+        Bottle* connections = v.asList();
+
+        mutex.lock();
+        for (std::multimap<ConstString,NodeItem>::iterator it = by_part_name.begin(); it != by_part_name.end(); it++) {
+            NodeItem& item = it->second;
+            if (!(item.isSubscriber() || item.isPublisher())) {
+                continue;
+            }
+            item.update();
+            item.contactable->getReport(report);
+        }
+        mutex.unlock();
```

Comment:
Not sure whether the following two `for` loops should also be protected by the mutex. I guess so, otherwise the `report` might change while iterating over it.